File: scripts/phase14_c2/muon.py
"""Phase 14 C2 — Muon optimizer for PyTorch (LoRA-adapter use case).

Mirror of nanogpt-rs/src/muon.rs. Same Newton-Schulz iteration
schedule (4 fast + 1 stabilize), same per-parameter dispatch (2-D
matrices → NS orthogonalize, 1-D → SGD-momentum fallback),
decoupled weight decay.

Designed for PEFT LoRA adapters: q_proj/v_proj LoRA delta_A
(in_dim × r) and delta_B (r × out_dim) are exactly the small-dense
2-D matrices Muon's NS targets.
"""

from typing import Iterable

import torch


def newton_schulz(x: torch.Tensor, n_steps: int = 5) -> torch.Tensor:
    """Approximate the orthogonal polar factor of `x` (X = Q · S → Q).

    Pre-normalize by Frobenius norm so σ_max ≤ 1, then 5 NS iterations
    with DeepSeek V4's coefficient schedule (4 fast + 1 stabilize).
    """
    if x.dim() != 2:
        raise ValueError(f"newton_schulz expects 2-D matrix, got {x.dim()}-D")
    frob = x.norm().clamp(min=1e-7)
    x = x / frob
    stage1 = (3.4445, -4.7750, 2.0315)
    stage2 = (2.0, -1.5, 0.5)
    n_stage1 = max(0, n_steps - 1)
    for i in range(n_steps):
        a, b, c = stage1 if i < n_stage1 else stage2
        xxt = x @ x.T
        xxt_x = xxt @ x
        xxt2_x = xxt @ xxt_x
        x = a * x + b * xxt_x + c * xxt2_x
    return x


class Muon(torch.optim.Optimizer):
    """Newton-Schulz orthogonalized SGD-momentum.

    For 2-D parameters: gradient → momentum → NS orthogonalize → step.
    For 1-D parameters: gradient → momentum → step (no NS).
    Decoupled weight decay (AdamW-style).

    Drop-in for `torch.optim.AdamW` for LoRA adapter training.
    """

    def __init__(
        self,
        params: Iterable,
        lr: float = 1e-3,
        momentum: float = 0.95,
        fallback_momentum: float = 0.9,
        weight_decay: float = 0.01,
        ns_steps: int = 5,
    ):
        if lr <= 0:
            raise ValueError(f"lr must be > 0, got {lr}")
        if not 0 <= momentum < 1:
            raise ValueError(f"momentum must be in [0, 1), got {momentum}")
        defaults = dict(
            lr=lr, momentum=momentum, fallback_momentum=fallback_momentum,
            weight_decay=weight_decay, ns_steps=ns_steps,
        )
        super().__init__(params, defaults)

    @torch.no_grad()
    def step(self, closure=None):
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for group in self.param_groups:
            lr = group["lr"]
            wd = group["weight_decay"]
            mom = group["momentum"]
            mom_1d = group["fallback_momentum"]
            ns_steps = group["ns_steps"]
            for p in group["params"]:
                if p.grad is None:
                    continue
                grad = p.grad
                state = self.state[p]
                if "momentum_buf" not in state:
                    state["momentum_buf"] = torch.zeros_like(p)
                m = state["momentum_buf"]
                # momentum update (β depends on rank — 2-D params get
                # the higher Muon momentum; 1-D fall back).
                beta = mom if p.dim() == 2 else mom_1d
                m.mul_(beta).add_(grad)

                # 2-D: NS orthogonalize the momentum direction.
                # 1-D: use raw momentum (SGD-mom fallback).
                if p.dim() == 2:
                    # Cast to fp32 for the NS computation; LoRA adapters
                    # may live in fp16.
                    direction = newton_schulz(m.float(), ns_steps).to(m.dtype)
                else:
                    direction = m

                # Decoupled weight decay (AdamW-style):
                #   θ ← θ · (1 − lr·wd) − lr · direction
                p.data.mul_(1.0 - lr * wd).add_(direction, alpha=-lr)

        return loss


# ---- Self-tests (run as `python muon.py`) ----
if __name__ == "__main__":
    torch.manual_seed(0)

    # 1. NS on identity → identity
    x = torch.eye(4)
    q = newton_schulz(x, 5)
    assert torch.allclose(q, x, atol=0.1), f"NS(I) ≠ I: {q}"
    print("[ok] newton_schulz(I) ≈ I")

    # 2. NS produces near-orthogonal columns
    x = torch.randn(6, 6)
    q = newton_schulz(x, 5)
    qtq = q.T @ q
    err = (qtq - torch.eye(6)).abs().max().item()
    assert err < 0.15, f"Q^T Q far from I: max_err={err}"
    print(f"[ok] newton_schulz random Q^T Q ≈ I (max_err {err:.4f})")

    # 3. Muon optimizer reduces a quadratic loss
    target = torch.randn(4, 4) * 1.0  # bigger target so loss starts higher
    w = torch.zeros(4, 4, requires_grad=True)
    opt = Muon([w], lr=0.1)
    losses = []
    for _ in range(80):
        loss = ((w - target) ** 2).mean()
        opt.zero_grad()
        loss.backward()
        opt.step()
        losses.append(loss.item())
    assert losses[-1] < losses[0] * 0.5, (
        f"Muon failed to reduce loss: start={losses[0]:.4f} end={losses[-1]:.4f}"
    )
    print(f"[ok] Muon: loss {losses[0]:.4f} → {losses[-1]:.4f}")

    # 4. 1-D fallback works
    target_1d = torch.tensor([1.0, 2.0, 3.0, 4.0])
    b = torch.zeros(4, requires_grad=True)
    opt = Muon([b], lr=0.05)
    for _ in range(40):
        loss = ((b - target_1d) ** 2).mean()
        opt.zero_grad()
        loss.backward()
        opt.step()
    assert (b.detach() - target_1d).abs().max() < 0.5, (
        f"1-D fallback failed: {b.detach()}"
    )
    print(f"[ok] 1-D fallback: b ≈ target ({b.detach().tolist()})")

    print("\nall muon.py self-tests passed")

File: nanogpt-rs/src/muon.rs
//! Phase 12 S1 — Muon optimizer (gradient orthogonalization).
//!
//! The Muon optimizer (Jordan et al. 2024, used in DeepSeek V4 2026)
//! is a variant of SGD with momentum where the *momentum tensor* is
//! orthogonalized via Newton-Schulz iterations before the parameter
//! update. The orthogonalization removes correlations between
//! gradient directions, giving faster convergence and better
//! stability than AdamW at large scale (DeepSeek used Muon for the
//! majority of modules in V4-Pro at 1.6T params).
//!
//! ## Per-parameter dispatch
//!
//! Muon's orthogonalization step only makes sense for 2-D weight
//! matrices. 1-D parameters (LayerNorm scales, biases, embedding
//! row vectors) fall back to standard SGD-with-momentum:
//!
//!   2-D `W ∈ ℝ^{m×n}`: g → momentum → NS-orthogonalize → step
//!   1-D `b ∈ ℝ^n`:     g → momentum → step (no NS)
//!
//! Following DeepSeek's hybrid pattern: Muon is the *primary*
//! optimizer; AdamW handles the small minority of 1-D params. We
//! simplify here by using SGD-momentum for those, which suffices
//! for our 50M-class models.
//!
//! ## Newton-Schulz iteration
//!
//! For an input matrix `X ∈ ℝ^{m×n}`, NS computes an approximation of
//! the orthogonal polar factor `Q` of `X` (i.e. `X = Q · S` with
//! `Q^T Q = I`). The iteration:
//!
//!   X_{k+1} = a · X_k + b · (X_k X_k^T) X_k + c · (X_k X_k^T)^2 X_k
//!
//! With DeepSeek V4's coefficient schedule:
//!
//!   stage 1 (4 iters): a, b, c = 3.4445, -4.7750, 2.0315 (fast convergence)
//!   stage 2 (1 iter):  a, b, c = 2.0,    -1.5,    0.5     (stabilize σ ≈ 1)
//!
//! Total 5 iterations. Pre-normalize `X` by spectral norm before NS
//! so all singular values start in `[0, 1]`, where the iteration is
//! well-behaved.
//!
//! ## Use from training
//!
//! ```rust,ignore
//! let opt = Muon::new(varmap.all_vars(), MuonConfig::default())?;
//! opt.backward_step(&loss)?;
//! ```
//!
//! Drop-in replacement for `AdamW::new(...)`. Same `Optimizer`
//! trait, same `set_learning_rate` / `backward_step` API.

use candle_core::{Result as CResult, Tensor, Var};
use candle_nn::Optimizer;

#[derive(Clone, Debug)]
pub struct MuonConfig {
    pub lr: f64,
    pub momentum: f64,
    pub weight_decay: f64,
    /// Number of Newton-Schulz iterations. DeepSeek V4 uses 5 (4
    /// fast-stage + 1 stabilize-stage).
    pub ns_steps: usize,
    /// Adam-like fallback hyperparameters for 1-D parameters where NS
    /// doesn't apply. We use the same `lr` and `weight_decay` as the
    /// 2-D path; these are momentum/eps for the SGD-mom fallback.
    pub fallback_momentum: f64,
}

impl Default for MuonConfig {
    fn default() -> Self {
        Self {
            lr: 1e-3,
            momentum: 0.95,
            weight_decay: 0.01,
            ns_steps: 5,
            fallback_momentum: 0.9,
        }
    }
}

#[derive(Debug)]
struct VarMuon {
    var: Var,
    momentum_buf: Var,
    is_matrix: bool,
}

#[derive(Debug)]
pub struct Muon {
    vars: Vec<VarMuon>,
    cfg: MuonConfig,
    step_t: usize,
}

impl Optimizer for Muon {
    type Config = MuonConfig;

    fn new(vars: Vec<Var>, cfg: MuonConfig) -> CResult<Self> {
        let vars = vars
            .into_iter()
            .filter(|var| var.dtype().is_float())
            .map(|var| {
                let shape = var.shape();
                let dtype = var.dtype();
                let device = var.device();
                let momentum_buf = Var::zeros(shape, dtype, device)?;
                let is_matrix = var.dims().len() == 2;
                Ok(VarMuon {
                    var,
                    momentum_buf,
                    is_matrix,
                })
            })
            .collect::<CResult<Vec<_>>>()?;
        Ok(Self {
            vars,
            cfg,
            step_t: 0,
        })
    }

    fn learning_rate(&self) -> f64 {
        self.cfg.lr
    }

    fn set_learning_rate(&mut self, lr: f64) {
        self.cfg.lr = lr;
    }

    fn step(&mut self, grads: &candle_core::backprop::GradStore) -> CResult<()> {
        self.step_t += 1;
        let lr = self.cfg.lr;
        let wd = self.cfg.weight_decay;
        for v in &self.vars {
            let Some(g) = grads.get(&v.var) else {
                continue;
            };
            // momentum update: m ← β·m + g
            let beta = if v.is_matrix {
                self.cfg.momentum
            } else {
                self.cfg.fallback_momentum
            };
            let new_m = ((v.momentum_buf.as_tensor() * beta)? + g)?;
            v.momentum_buf.set(&new_m)?;

            // For 2-D matrices: orthogonalize the momentum (this is the
            // Muon step). For 1-D: skip NS, use raw momentum (SGD-mom).
            let direction = if v.is_matrix {
                newton_schulz(&new_m, self.cfg.ns_steps)?
            } else {
                new_m.clone()
            };

            // weight decay (decoupled, AdamW-style):
            //   θ ← θ · (1 − lr·wd) − lr · direction
            let theta = v.var.as_tensor();
            let decayed = (theta * (1.0 - lr * wd))?;
            let new_theta = (decayed - (direction * lr)?)?;
            v.var.set(&new_theta)?;
        }
        Ok(())
    }
}

/// Newton-Schulz polar-factor iteration. Approximates the orthogonal
/// matrix `Q` such that `X = Q·S` (polar decomposition). Returns the
/// orthogonalized version of `x` (same shape).
///
/// Pre-normalizes by spectral norm so every singular value starts in
/// `[0, 1]`. With DeepSeek V4's coefficient schedule (4 fast steps +
/// 1 stabilize), 5 total iterations bring singular values to ≈ 1.
///
/// Numerical safety: clamps the spectral-norm pre-divisor to ≥ 1e-7.
pub fn newton_schulz(x: &Tensor, n_steps: usize) -> CResult<Tensor> {
    let dims = x.dims();
    if dims.len() != 2 {
        candle_core::bail!(
            "newton_schulz expects a 2-D matrix, got rank {}",
            dims.len()
        );
    }
    // Pre-normalize: scale so largest singular value ≤ 1. Frobenius
    // norm is an upper bound on σ_max, so dividing by it is safe.
    let frob = x.sqr()?.sum_all()?.sqrt()?.to_scalar::<f32>()? as f64;
    let denom = frob.max(1e-7);
    let mut x = (x / denom)?;

    // (m, n) → operate in the smaller side. If m < n we'd prefer to
    // work with X^T so the inner dimension is small. For our 50M
    // model the matrices are square-ish; keep simple form.
    let stage1 = (3.4445_f64, -4.7750, 2.0315);
    let stage2 = (2.0_f64, -1.5, 0.5);
    let n_stage1 = n_steps.saturating_sub(1);
    for i in 0..n_steps {
        let (a, b, c) = if i < n_stage1 { stage1 } else { stage2 };
        // X_{k+1} = a·X + b·X X^T X + c·(X X^T)² X
        let xxt = x.matmul(&x.t()?)?;
        let xxt_x = xxt.matmul(&x)?;
        let xxt2_x = xxt.matmul(&xxt_x)?;
        x = ((x * a)? + (xxt_x * b)? + (xxt2_x * c)?)?;
    }
    Ok(x)
}

#[cfg(test)]
mod tests {
    use super::*;
    use candle_core::{DType, Device, Tensor};
    use candle_nn::{Module, VarBuilder, VarMap};

    #[test]
    fn newton_schulz_identity_returns_near_identity() {
        // NS on the identity matrix should keep it the identity (its
        // own polar factor). Small numerical drift is acceptable.
        let device = Device::Cpu;
        let id = Tensor::from_vec(vec![1.0_f32, 0.0, 0.0, 1.0], (2, 2), &device).unwrap();
        let out = newton_schulz(&id, 5).unwrap();
        let out_v = out.to_vec2::<f32>().unwrap();
        // Off-diagonals near 0, diagonals near 1.
        assert!((out_v[0][0] - 1.0).abs() < 0.05);
        assert!(out_v[0][1].abs() < 0.05);
        assert!(out_v[1][0].abs() < 0.05);
        assert!((out_v[1][1] - 1.0).abs() < 0.05);
    }

    #[test]
    fn newton_schulz_gives_orthogonal_columns() {
        // After NS, X^T X should be near identity (columns orthonormal).
        //
        // Fixed matrix, not `randn`. Unseeded this test failed about 1 run in
        // 6: a random 4x4 Gaussian is occasionally near-singular, and 5
        // Newton-Schulz iterations do not bring such a matrix inside the 0.1
        // tolerance. That made it assert "NS orthogonalizes ANY random matrix
        // in 5 steps", which is false and not what it is here to check — and
        // CI runs `cargo test --workspace` as a strict gate, so it was a ~17%
        // chance of a red build for no reason. `Device::Cpu` rejects
        // `set_seed` ("cannot seed the CPU rng"), so the fix is a literal
        // well-conditioned matrix rather than a seeded draw.
        let device = Device::Cpu;
        let x = Tensor::from_vec(
            vec![
                1.2_f32, -0.4, 0.3, 0.1, //
                0.5, 1.4, -0.2, 0.6, //
                -0.3, 0.2, 1.1, -0.5, //
                0.4, -0.6, 0.7, 1.3,
            ],
            (4, 4),
            &device,
        )
        .unwrap();
        let q = newton_schulz(&x, 5).unwrap();
        // Q^T Q ≈ I_4
        let qtq = q.t().unwrap().matmul(&q).unwrap();
        let qtq_v = qtq.to_vec2::<f32>().unwrap();
        for (i, row) in qtq_v.iter().enumerate().take(4) {
            for (j, val) in row.iter().enumerate().take(4) {
                let target = if i == j { 1.0_f32 } else { 0.0 };
                assert!(
                    (val - target).abs() < 0.1,
                    "Q^T Q[{i}][{j}]={val} expected {target}"
                );
            }
        }
    }

    #[test]
    fn newton_schulz_rejects_non_2d() {
        let device = Device::Cpu;
        let v = Tensor::randn(0_f32, 1.0, 5, &device).unwrap();
        assert!(newton_schulz(&v, 5).is_err());
        let cube = Tensor::randn(0_f32, 1.0, (2, 3, 4), &device).unwrap();
        assert!(newton_schulz(&cube, 5).is_err());
    }

    #[test]
    fn muon_step_decreases_quadratic_loss() {
        // Train W ∈ ℝ^{4×4} to minimize ||W − target||² via Muon.
        // After a few steps, the loss should decrease.
        let device = Device::Cpu;
        let varmap = VarMap::new();
        let vb = VarBuilder::from_varmap(&varmap, DType::F32, &device);
        let lin = candle_nn::linear_no_bias(4, 4, vb).unwrap();

        let target = Tensor::from_vec(
            (0..16).map(|i| i as f32 * 0.1).collect::<Vec<_>>(),
            (4, 4),
            &device,
        )
        .unwrap();

        let mut opt = Muon::new(
            varmap.all_vars(),
            MuonConfig {
                lr: 1e-1,
                ..MuonConfig::default()
            },
        )
        .unwrap();

        // Use the linear layer to drive a loss against the target.
        let x = Tensor::randn(0_f32, 1.0, (8, 4), &device).unwrap();
        let target_y = x.matmul(&target).unwrap();
        let loss0 = (lin.forward(&x).unwrap() - &target_y)
            .unwrap()
            .sqr()
            .unwrap()
            .mean_all()
            .unwrap()
            .to_scalar::<f32>()
            .unwrap();
        for _ in 0..30 {
            let pred = lin.forward(&x).unwrap();
            let loss = (pred - &target_y)
                .unwrap()
                .sqr()
                .unwrap()
                .mean_all()
                .unwrap();
            opt.backward_step(&loss).unwrap();
        }
        let loss1 = (lin.forward(&x).unwrap() - &target_y)
            .unwrap()
            .sqr()
            .unwrap()
            .mean_all()
            .unwrap()
            .to_scalar::<f32>()
            .unwrap();
        assert!(
            loss1 < loss0 * 0.5,
            "Muon should at least halve loss; loss0={loss0} loss1={loss1}"
        );
    }

    #[test]
    fn muon_set_learning_rate_round_trips() {
        let device = Device::Cpu;
        let varmap = VarMap::new();
        let _vb = VarBuilder::from_varmap(&varmap, DType::F32, &device);
        let mut opt = Muon::new(varmap.all_vars(), MuonConfig::default()).unwrap();
        opt.set_learning_rate(5e-4);
        assert!((opt.learning_rate() - 5e-4).abs() < 1e-12);
    }

    #[test]
    fn muon_handles_1d_parameters_via_sgd_momentum_fallback() {
        // 1-D bias vector: NS should NOT be invoked (bail otherwise).
        // Loss = mean((b − target)²). Muon should still drive b → target.
        let device = Device::Cpu;
        let varmap = VarMap::new();
        let target = Tensor::from_vec(vec![1.0_f32, 2.0, 3.0, 4.0], 4, &device).unwrap();
        let b = candle_core::Var::zeros(4, DType::F32, &device).unwrap();
        // Manually register in varmap so Muon picks it up.
        // Easiest: build a varmap with a custom var and pass via from_slice.
        let mut opt = Muon::from_slice(
            &[&b],
            MuonConfig {
                lr: 5e-2,
                ..MuonConfig::default()
            },
        )
        .unwrap();
        let _ = varmap;
        for _ in 0..40 {
            let loss = (b.as_tensor() - &target)
                .unwrap()
                .sqr()
                .unwrap()
                .mean_all()
                .unwrap();
            opt.backward_step(&loss).unwrap();
        }
        let final_v = b.as_tensor().to_vec1::<f32>().unwrap();
        for (got, want) in final_v.iter().zip([1.0_f32, 2.0, 3.0, 4.0].iter()) {
            assert!(
                (got - want).abs() < 0.5,
                "1-D bias should converge toward target; got {final_v:?}"
            );
        }
    }
}
